lib/luadata.h

```diff
@@ -17,6 +17,44 @@ LUNATIK_LIB(data);
 
 #define luadata_clear(o)	(luadata_reset((o), NULL, 0, LUADATA_OPT_KEEP))
 
+typedef struct luadata_s {
+    char *ptr;
+    size_t size;
+    uint8_t opt;
+} luadata_t;
+
+LUNATIK_PRIVATECHECKER(luadata_check, luadata_t *);
+
+/***
+ * Bounds-checked pointer calculation. Returns pointer on success, raises Lua error on failure.
+ * @param L Lua state
+ * @param ix Argument index for error reporting
+ * @param data luadata object
+ * @param offset Byte offset
+ * @param length Access length
+ * @return Valid pointer within bounds
+ */
+static inline void *luadata_checkbounds(lua_State *L, int ix, luadata_t *data, lua_Integer offset, lua_Integer length)
+{
+	int bounds = offset >= 0 && length > 0 && offset + length <= data->size;
+	luaL_argcheck(L, bounds, ix, "out of bounds");
+	return (void *)(data->ptr + offset);
+}
+
+static inline void *luadata_checkrange(lua_State *L, int idata, int ioffset, int ilen, size_t *len_out)
```

Comment:
I think we should pass just one index on the stack and assume them sequential.. like a CFunction..